### PR TITLE
Add method to retrieve full tracking graph

### LINKED ISSUE #227: Add method to retreive full tracking graph
+ [x] All nodes
+ [x] All edges and weights from Bayesian updates
+ [x] All edges from ILP hypotheses
+ [x] All hyperedges from ILP hypotheses

### PR DESCRIPTION
**Pull request by quantumjot**
This adds a method to enable retrieval of the full track graph from btrack.  

* Internally, this means that a flag is set to store the edges and hyperedges in the `TrackManager` class in the C++ lib.
* I've added a new configuration option `TrackerConfig.store_candidate_graph` which lets the `TrackManager` know to reserve space and store the candidate graph edges.
* In Python, These can be returned with the `candidate_graph_edges()` method:

```python
with btrack.BayesianTracker() as tracker:
    tracker.configuration.store_candidate_graph = True
    ...
    candidate_graph = tracker.candidate_graph_edges()
```
* I've also added a couple of tests (`test_tracker_candidate_graph`) to make sure that this function returns edges when the config flag is set.
* This also tests the CI based compilation introduced in #231
* I also applied some automated formatting (Google style) to the C++ code and applied a few small changes to supress warnings about definitions in the headers.

Closes #227

### CODE REVIEW

**Comment by paddyroddy** [comment]
> Tests will fail until #231 is merged.

It's merged now

File: btrack/core.py [hunk 1]
```diff
@@ -685,7 +693,24 @@ def to_napari(
...
+            )
+        graph_edges = [
+            self._lib.get_graph_edge(self._engine, idx)
+            for idx in range(num_edges)
+        ]
+        return graph_edges
+
+
+if __name__ == "__main__":
+    pass
```
Comment (paddyroddy) [inline btrack/core.py:716]: Why add this?

File: btrack/include/bayes.h [hunk 2]
```diff
@@ -21,19 +21,12 @@
...
-    double likelihood
-  );
-}
+namespace BayesianUpdateFunctions {
+std::tuple<double, double> safe_bayesian_update(double prior_assign,
+                                                double prob_assign,
+                                                double prob_not_assign);
 
+double safe_bayesian_update_simple(double prior, double likelihood);
+}  // namespace BayesianUpdateFunctions
```
Comment (paddyroddy) [inline btrack/include/bayes.h:30]: Is the comment needed?

File: btrack/include/defs.h [hunk 3]
```diff
@@ -17,13 +17,37 @@
...
+#ifndef VERSION_MINOR
+#define VERSION_MINOR 0
+#endif
+#ifndef VERSION_BUILD
+#define VERSION_BUILD 0
+#endif
+
+#ifndef DEBUG
+#define DEBUG 0
+#endif
```
Comment (paddyroddy) [inline btrack/include/defs.h:40]: I had found this was required in #230 & #225

File: tests/test_tracker.py [hunk 4]
```diff
@@ -94,3 +95,22 @@ def test_tracker_frames():
...
+    # tracker.store_candidate_graph = True
+    assert tracker.store_candidate_graph == store_graph
+
+    edges = tracker.candidate_graph_edges()
+
+    # graph should contain edges
+    if store_graph:
+        assert edges, f"Found {len(edges)} edges in candidate graph."
+    else:
+        assert not edges, f"Found {len(edges)} edges in candidate graph."
```
Comment (paddyroddy) [inline tests/test_tracker.py:116]: I believe it's bad practice to have an `if` in a test. This should maybe be two separate tests

**Review by paddyroddy — CHANGES_REQUESTED** [review]
I stopped looking at the C++ as it's huge and mainly formatting/hard to differentiate.

One thought is that we should add a formatting pre-commit hook for the C++, e.g. https://github.com/pre-commit/mirrors-clang-format or https://github.com/pocc/pre-commit-hooks

File: btrack/core.py (hunk 1 above)
Comment (quantumjot) [reply btrack/core.py]: dang it, I think that was from the original branch. removed

File: btrack/include/defs.h [hunk 5]
```diff
@@ -17,13 +17,37 @@
...
+
+// perform strict tracklet linking, i.e. terminate a track if there are any
+// perceived conflicts
+#ifndef STRICT_TRACKLET_LINKING
+#define STRICT_TRACKLET_LINKING 0
+#endif
+
+// define a filepath for debugging output
+#define DEBUG_FILEPATH "/home/quantumjot/Documents/"
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply btrack/include/defs.h]: Done. Now creates a temporary directory

File: tests/test_tracker.py (hunk 4 above)
Comment (quantumjot) [reply tests/test_tracker.py]: I found out that you can just do `bool(some_list)` that will evaluate as True if it contains something, so I could replace this if statement with a comparison

**Comment by quantumjot** [comment]
> One thought is that we should add a formatting pre-commit hook for the C++, e.g. https://github.com/pre-commit/mirrors-clang-format or https://github.com/pocc/pre-commit-hooks

I agree, this would be a good idea.

**Comment by quantumjot** [comment]
Thanks @paddyroddy. Suggestions have been implemented now.

**Pull request merged**